ports/raspberrypi/common-hal/digitalio/DigitalInOut.c

```diff
@@ -110,9 +116,6 @@ digitalinout_result_t common_hal_digitalio_digitalinout_set_drive_mode(
     const uint8_t pin = self->pin->number;
     bool value = common_hal_digitalio_digitalinout_get_value(self);
     self->open_drain = drive_mode == DRIVE_MODE_OPEN_DRAIN;
-    if (self->open_drain) {
-        gpio_put(pin, false);
```

Comment:
What is wrong with this? Shouldn't the output value be false but toggle output enable?

Reply:
Suppose `value` is true. This will pull down the output temporarily, and then it will got back up on line 119/122.

Reply:
See https://github.com/adafruit/circuitpython/blob/d4bf0d5e2d16611434a955ca3072da4626dc0038/ports/atmel-samd/common-hal/digitalio/DigitalInOut.c#L132-L143.

Reply:
In this implementation we don't set the value in `set_value` at all when in open drain mode. So, we need to set it to false somewhere. Otherwise you'll just toggle asserting high. (I think you can move this bit to after the `set_value` call below because it'll turn off output enable when value is high.)

Reply:
OK, I've gone over this carefully, and tested with `bitbangio.I2C`. I think it's correct now.

`common_hal_digitalio_digitalinout_set_drive_mode()` is now identical to the `atmel-samd` implementation. The difference between the two implementations is that the `atmel-samd` impl infers the state of the pin in `get_value()` if it's open-drain by looking at the direction (so it doesn't need to actually set the value ever in that case). In the `raspberrypi` implementation, we always set the register value, so we use that instead.

`set_value()` takes care of switching the direction as necessary. It is also careful to do the direction switch and the pin-value setting in the correct order to avoid any small glitches. The order is reversed depending on the value.